### PR TITLE
Support for absolute paths & module handles in EAT hooks

### PR DESCRIPTION
**Pull request by acidicoala**
This PR implements support for absolute module paths & module handles for EAT Hooks as discussed in #114.

Some miscellaneous optimizations suggested by Clang-tidy have also been applied.

New tests have been added that cover the new use cases, and test names have been rephrased to reflect that change.

### CODE REVIEW

File: UnitTests/windows/TestEatHook.cpp [hunk 1]
```diff
@@ -4,44 +4,45 @@
...
 #include "polyhook2/PolyHookOsIncludes.hpp"
+#include "polyhook2/Detour/ADetour.hpp"
 
 EffectTracker eatEffectTracker;
 
 typedef void(* tEatTestExport)();
 uint64_t oEatTestExport;
 
-extern "C" __declspec(dllexport) NOINLINE void EatTestExport()
+extern "C" [[maybe_unused]] __declspec(dllexport) NOINLINE void EatTestExport()
```
Comment (stevemk14ebr) [inline UnitTests/windows/TestEatHook.cpp:14]: Please remove this function attribute I do not prefer to have these regardless of warnings.

File: polyhook2/PE/EatHook.hpp [hunk 2]
```diff
@@ -16,7 +16,9 @@ namespace PLH {
 class EatHook : public IHook {
 public:
 	EatHook(const std::string& apiName, const std::wstring& moduleName, const char* fnCallback, uint64_t* userOrigVar);
-	EatHook(const std::string& apiName, const std::wstring& moduleName, const uint64_t fnCallback, uint64_t* userOrigVar);
+	EatHook(const std::string& apiName, const std::wstring& moduleName, uint64_t fnCallback, uint64_t* userOrigVar);
```
Comment (stevemk14ebr) [inline polyhook2/PE/EatHook.hpp:19]: const on the fnCallback argument please

File: polyhook2/PE/EatHook.hpp [hunk 3]
```diff
@@ -31,10 +33,15 @@ class EatHook : public IHook {
 	virtual HookType getType() const override {
 		return HookType::EAT;
 	}
-private:
+
+protected:
```
Comment (stevemk14ebr) [inline polyhook2/PE/EatHook.hpp:37]: why the change to protected? I wouldn't expect this to ever be a parent class really, do you see a case someone may want to derive from this as a base class?

File: sources/EatHook.cpp [hunk 4]
```diff
@@ -76,53 +88,69 @@ bool PLH::EatHook::unHook() {
...
+	auto* ldr = (PPEB_LDR_DATA)peb->Ldr;
+	auto* dte = (LDR_DATA_TABLE_ENTRY*)ldr->InLoadOrderModuleList.Flink;
 
-	// find loaded module from peb
-	for (auto* dte = (LDR_DATA_TABLE_ENTRY*)ldr->InLoadOrderModuleList.Flink;
-        dte->DllBase != NULL;
-        dte = (LDR_DATA_TABLE_ENTRY*)dte->InLoadOrderLinks.Flink) {
+	// Empty module name implies current process
+	if(m_moduleName.empty()){
+		return reinterpret_cast<uint64_t>(dte->DllBase);
```
Comment (stevemk14ebr) [inline sources/EatHook.cpp:116]: Please revert all C++ casts to the C style.

**Review by stevemk14ebr — CHANGES_REQUESTED** [review]

**Comment by stevemk14ebr** [comment]
I think this looks functionally correct, thanks for working on this! I am not a fan of some of the style changes and I've listed the ones I'd like reverted, otherwise looks great thanks!

File: UnitTests/windows/TestEatHook.cpp (hunk 1 above)
Comment (acidicoala) [reply UnitTests/windows/TestEatHook.cpp:14]: Sure

File: polyhook2/PE/EatHook.hpp (hunk 2 above)
Comment (acidicoala) [reply polyhook2/PE/EatHook.hpp:19]: But it has no effect on semantics/logic, afaik. A primitive value is always copied, so the argument is always constant from the caller side. It only makes a difference in function body, which is what I have done in the source file.

See: https://stackoverflow.com/a/46292715

File: polyhook2/PE/EatHook.hpp (hunk 3 above)
Comment (acidicoala) [reply polyhook2/PE/EatHook.hpp:37]: In general, I consider it a sub-optimal design choice to set any class members to private for library classes, because if there is going to be a need to modify the behavior for unforeseeable reasons, the consumer of the library won't be cursing the library author(s) for needlessly restricting access. It's not a pleasant experience to maintain a private fork of the library because it is impossible to inherit a library class to modify it. But that's just my opinion, and I wanted to hear your view on this issue as well.

File: sources/EatHook.cpp (hunk 4 above)
Comment (acidicoala) [reply sources/EatHook.cpp:116]: Sure.

**Comment by acidicoala** [comment]
@stevemk14ebr. I don't know if you receive notifications for comment replies, but I left some questions just to understand on your perspective on these styling conventions, before I implement your suggestions. Thank you!

File: polyhook2/PE/EatHook.hpp (hunk 2 above)
Comment (stevemk14ebr) [reply polyhook2/PE/EatHook.hpp:19]: you are correct, but it's helpful to see the interface as a user that this value is const and it can prevent errors in constructors especially when you may accidentally assign to the parameter. This is more style - as most sprinkling on const is - though there are benefits.

File: polyhook2/PE/EatHook.hpp (hunk 3 above)
Comment (stevemk14ebr) [reply polyhook2/PE/EatHook.hpp:37]: That's reasonable justification honestly. I set classes I expected to be derived from to protected already, I just never expected a case it would be useful for this but it doesn't hurt so why not.

**Comment by stevemk14ebr** [comment]
I can see them! Responded

**Comment by stevemk14ebr** [comment]
thanks!

**Comment by stevemk14ebr** [comment]
I am open to all classes the derive from IHook to be widened to protected yea

**Comment by acidicoala** [comment]
Cool, will submit a PR for that later!
